Kuadrant/kcp-glbc
# Custom domains verification

Closes #80 

## Verification steps
1. To mock the TXT verification, uncomment the following block, to make use the ConfigMap resolver:  https://github.com/Kuadrant/kcp-glbc/blob/3f8124e5c4ce84b9ca61226b01ab9b5de34748e8/cmd/kcp-glbc/main.go#L265-L269
2. Start KCP
    ```sh
    make local-setup
    ```

3. In another terminal, run the controller
    ```sh
    export KUBECONFIG=config/deploy/local/kcp.kubeconfig
    # If mocking the DNS management:
    # export AWS_DNS_PUBLIC_ZONE_ID=FAKE_ZONE_ID
    ./bin/kubectl-kcp workspace use root:default:kcp-glbc
    ./bin/kcp-glbc --kubeconfig .kcp/admin.kubeconfig --context system:admin --enable-custom-hosts
    ```
4. If mocking the TXT verification, in another terminal, create the mock ConfigMap
    ```sh
    export KUBECONFIG=config/deploy/local/kcp.kubeconfig
    ./bin/kubectl-kcp workspace use root:default:kcp-glbc
    kubectl apply -f samples/domainverfication/hosts.yaml
    ```
5. In another terminal, create the Ingress, DomainVerification and workload
    ```sh
    export KUBECONFIG=.kcp/admin.kubeconfig
    ./bin/kubectl-kcp workspace use root:default:kcp-glbc-user
    kubectl apply -f samples/domainverification/example.yaml
    ```
6. Verify that the DomainVerification is reconciled, remains unverified and the token is set
7. Verify that the Ingress is reconciled and
    1. The custom host is replaced for the generated one
    2. The custom host rule is preserved in the `kuadrant.dev/custom-hosts.pending` annotation of the Ingress
8. Test the domain verification. Take the token from the `DomainVerification` resource and set it in the TXT entry
    > If mocking the TXT resolution, set it in the `hosts` ConfigMap
9. Verify that
    1. The DomainVerification resource is reconciled and marked as verified
    2. The Ingress resource is reconciled and the custom host is added back to the `spec` 

#### To verify the empty host scenario

1. Edit the Ingress, appending a rule with the empty host:
    ```yaml
    - host: ""
      http:
        paths:
        - backend:
            service:
              name: httpecho-both
              port:
                number: 80
          path: /mypath
          pathType: Prefix
    ```
2. Verify that the ingress is reconciled and the empty host is replaced by the generated host
